lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +257,358 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+
+    DI.recalculate(*F);
+    PDI.recalculate(*F);
+  }
+  return changed;
+}
+
+namespace {
+
+/// Appends the given arguments to the given edge. Deletes the old TermInst
+/// and returns a new TermInst with the appropriate number of args.
+TermInst* appendArguments(TermInst *termInst, SILBasicBlock *target,
+                     ArrayRef<SILValue> newArgs) {
+
+  SILBuilder builder(termInst);
+  TermInst *newTermInst = nullptr;
+
+  auto createArgsArray = [target,
+                          &newArgs](const OperandValueArrayRef &currentArgs,
+                                    bool appendNewArgs) {
+    SmallVector<SILValue, 8> args(currentArgs.begin(), currentArgs.end());
+    if (appendNewArgs) {
+      args.append(newArgs.begin(), newArgs.end());
+      assert(args.size() == target->getNumArguments() &&
+             "Number of final arguments does not match target's arguments.");
+    }
+    return args;
+  };
+  if (auto *branch = dyn_cast<BranchInst>(termInst)) {
+    assert(branch->getDestBB() == target &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> args =
+        createArgsArray(branch->getArgs(), /*appendNewArgs=*/true);
+    newTermInst =
+        builder.createBranch(branch->getLoc(), branch->getDestBB(), args);
+  } else if (auto *condBranch = dyn_cast<CondBranchInst>(termInst)) {
+    bool isTrueEdge = condBranch->getTrueBB() == target;
+    assert(((isTrueEdge && condBranch->getTrueBB() == target) ||
+            (!isTrueEdge && condBranch->getFalseBB() == target)) &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> trueArgs = createArgsArray(
+        condBranch->getTrueArgs(), /*appendNewArgs=*/isTrueEdge);
+    SmallVector<SILValue, 8> falseArgs = createArgsArray(
+        condBranch->getFalseArgs(), /*appendNewArgs=*/!isTrueEdge);
+    newTermInst = builder.createCondBranch(
+        condBranch->getLoc(), condBranch->getCondition(),
+        condBranch->getTrueBB(), trueArgs, condBranch->getFalseBB(), falseArgs,
+        condBranch->getTrueBBCount(), condBranch->getFalseBBCount());
+  } else {
+    // At the moment we can only add arguments to br and cond_br.
+    llvm_unreachable("Can't add argument to terminator");
+  }
+  // Remove the old terminator instruction.
+  termInst->dropAllReferences();
+  termInst->eraseFromParent();
+  return newTermInst;
+}
+
+/// Creates a TF handle of appropriate type that is initialized to the given
+/// element value.
+SILValue createTFHandleFromElement(SILBuilder &builder, SILLocation location,
+                                   SILValue element) {
+  assert(builder.hasValidInsertionPoint());
+  ASTContext& context = builder.getASTContext();
+  SILType elementType = element->getType();
+  // Literals take attributes specifying the dtype, value, and device.
+  std::string opName("__tfop_Const,dtype$dtype,value$tensor,__device");
+  SmallVector<SILValue, 8> operands;
```

Comment:
There are exactly three operands, so there's no need for `SmallVector`. A C array `SILValue operands[]` would be simpler and more efficient.

Reply:
Done.